server/aesdsocket.c

```diff
@@ -54,51 +83,230 @@ void *get_in_addr(struct sockaddr *sa) {
+	char read_char;
+	char recv_buf[CHUNK_SIZE];
+	struct thread_cleanup_data cd;
+
+	// Cast paramater as correct struct
+	struct thread_data *t_data = (struct thread_data *) thread_param;
+
+	// Set up data for easy cleanup
+	recv_vec.buf = NULL;
+	send_vec.buf = NULL;
+	cd.t_data = t_data;
+	cd.recv_vec = &recv_vec;
+	cd.send_vec = &send_vec;
+
+	// Clear receive buffer to prepare for receive
+	if(vector_init(&recv_vec)){
+		syslog(LOG_ERR, "vec_init fail\n");
+		thread_cleanup(&cd);
+		return NULL;
+	}
+	
+	// Loop until client closes connection
+	while(!sig_received){
+		received = 0;
+
+		// Call receive until we've received a newline
+		while(!(new_line = vector_find(&recv_vec, recv_vec.len - received, '\n'))){
+
+			// Perform non-blocking receive
+			do {
+				received = recv(t_data->client_fd, recv_buf, CHUNK_SIZE, 0);
+			} while (received == -1 && errno == EAGAIN);
+
+			if(received < 0) {
+				syslog(LOG_ERR, "error on syscall: recv");
+				thread_cleanup(&cd);
+				return NULL;
+			}
+			else if(received > 0){
+				if(vector_append(&recv_vec, recv_buf, received)){
+					syslog(LOG_ERR, "vec_append fail\n");
+					thread_cleanup(&cd);
+					return NULL;
+				}
+			}
+			else {
+				// Connection was terminated, follow into the if below
+				// since double break isn't well defined
+				break;
+			}
+		}
+
+		// Will have fallen out of above loop
+		if(received == 0){
+			break;
+		}
+
+		// Write from the receive buffer into the data file one line at a time
+		written = 0;
+		while((new_line = vector_find(&recv_vec, written, '\n'))){
+			pthread_mutex_lock(&data_file.mtx);
+			if((rv = write(data_file.fd, recv_vec.buf+written, new_line + 1 - (char *)(recv_vec.buf+written))) == -1){
+				pthread_mutex_unlock(&data_file.mtx);
+				syslog(LOG_ERR, "error writing data to file");
+				thread_cleanup(&cd);
+				return NULL;				
+			}
+			pthread_mutex_unlock(&data_file.mtx);
+			written += rv;
+		}
+
+		// If we have have any extra data in the receive buffer, carry it through
+		// to the next packet, otherwise reset the buffer
+		if(written < recv_vec.len){
+			vector_carryover(&recv_vec, written);
+		}
+		else{
+			vector_close(&recv_vec);
+			vector_init(&recv_vec);
+		}
+
+		if(vector_init(&send_vec)){
+			syslog(LOG_ERR, "vec_init fail\n");
+			thread_cleanup(&cd);
+			return NULL;
+		}
+
+		pthread_mutex_lock(&data_file.mtx);
+		// Seek back to start of file for read
+		if(lseek(data_file.fd, 0, SEEK_SET) == -1){
+			pthread_mutex_unlock(&data_file.mtx);
+			syslog(LOG_ERR, "error seeking to start of file");
+			thread_cleanup(&cd);
+			return NULL;
+		}
+		pthread_mutex_unlock(&data_file.mtx);
+
+		// Loop through chars in file
+		pthread_mutex_lock(&data_file.mtx);
+		while(1){
+			if((rv = read(data_file.fd, &read_char, 1)) != 1){
+				break;
+			}
+
+			// Apeend each char to the line buffer
+			if(vector_append(&send_vec, &read_char, 1)){
+				syslog(LOG_ERR, "vec_append fail\n");
+				thread_cleanup(&cd);
+				return NULL;
```

Comment:
@jbohn3353 I think you are missing pthread_mutex_unlock in several of these cleanup return paths.  I'm not sure you need the locking over the entire while loop but I'd suggest a design where the mutex lock is held around a function call and cleanup is always performed on return regradless of success/failure